File: 1 - Theory of Computation/chpt1.nim
# EXERCISES

# Exercise 1.8: Newton's Method for cube root
proc cube_root(x:float, guess:float = 1.0, tolerance:float = 0.001): float = 
    proc close_enough(a:float): bool = abs(x - (a * a * a)) < tolerance
    proc improve_guess(a:float): float = ((x / (a * a)) + 2 * a) / 3

    proc newtons_method(a:float): float = 
        if close_enough(a):
            return a
        else:
            newtons_method(improve_guess(a))
    
    return newtons_method(guess)

echo cube_root(9)
echo cube_root(27)


# Exercise 1.11
proc f_recursive(n:int): int = 
    if n < 3:
        return n
    else:
        return f_recursive(n - 1) + 2*f_recursive(n - 2) + 3*f_recursive(n - 3)

echo f_recursive(14)

File: 1 - Theory of Computation/chpt1.jl
# EXERCISES
# Exercise 1.8

function cube_root(x, guess=1, tolerance=0.001)
    close_enough(a) = abs(x - a^3) < tolerance
    improve_guess(a) = ((x / a^2) + 2a) / 3

    function newtons_method(y)
        if (close_enough(y))
            return y
        else
            newtons_method(improve_guess(y))
        end
    end

    return newtons_method(guess)
end

println(cube_root(27))


# Exercise 1.11
function f_recursive(n)
    if n < 3
        return n
    else
        return f_recursive(n - 1) + 2f_recursive(n - 2) + 3f_recursive(n - 3)
    end
end

function f_iterative(n)

end

println(f_recursive(14))
println(f_iterative(14))